table/tables/mutation_checker.go

```diff
@@ -0,0 +1,244 @@
+func checkRowMutationWithData(sessVars *variable.SessionVars, mutationValue []byte, expectedData []types.Datum, columnMap map[int64]*model.ColumnInfo) error {
+	columnFieldMap := make(map[int64]*types.FieldType)
+	for id, col := range columnMap {
+		columnFieldMap[id] = &col.FieldType
+	}
+	decodedData, err := tablecodec.DecodeRowToDatumMap(mutationValue, columnFieldMap, sessVars.Location())
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// NOTE: we cannot check if the decoded values contain all columns since some columns may be skipped. It can even be empty
+	// Instead, we check that decoded index values are consistent with the input row.
+
+	for columnID, decodedDatum := range decodedData {
+		inputDatum := expectedData[columnMap[columnID].Offset]
+		cmp, err := decodedDatum.CompareDatum(sessVars.StmtCtx, &inputDatum)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if cmp != 0 {
+			logutil.BgLogger().Error("inconsistent row mutation", zap.String("decoded datum", decodedDatum.String()),
+				zap.String("input datum", inputDatum.String()))
+			return errors.New(fmt.Sprintf("inconsistent row mutation, row datum = {%v}, input datum = {%v}",
+				decodedDatum.String(), inputDatum.String()))
+		}
+	}
+
+	return nil
+}
+
+func collectTableMutationsFromBufferStage(t *TableCommon, memBuffer kv.MemBuffer, sh kv.StagingHandle) []mutation {
+	mutations := make([]mutation, 0)
+	inspector := func(key kv.Key, flags kv.KeyFlags, data []byte) {
+		// only check the current table
+		if tablecodec.DecodeTableID(key) == t.physicalTableID {
+			mutations = append(mutations, mutation{key, flags, data})
+		}
+	}
+	memBuffer.InspectStage(sh, inspector)
+	return mutations
+}
+
+// compareIndexData compares the decoded index data with the input data.
+// Returns error if the index data is not a subset of the input data.
+func compareIndexData(sc *stmtctx.StatementContext, cols []*table.Column, indexData, input []types.Datum, indexInfo *model.IndexInfo) error {
+	for i, decodedMutationDatum := range indexData {
+		expectedDatum := input[indexInfo.Columns[i].Offset]
+
+		tablecodec.TruncateIndexValue(&expectedDatum, indexInfo.Columns[i],
+			cols[indexInfo.Columns[i].Offset].ColumnInfo)
+		tablecodec.TruncateIndexValue(&decodedMutationDatum, indexInfo.Columns[i],
+			cols[indexInfo.Columns[i].Offset].ColumnInfo)
+
+		comparison, err := decodedMutationDatum.CompareDatum(sc, &expectedDatum)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if comparison != 0 {
+			logutil.BgLogger().Error("inconsistent index values",
+				zap.String("truncated mutation datum", fmt.Sprintf("%v", decodedMutationDatum)),
+				zap.String("truncated expected datum", fmt.Sprintf("%v", expectedDatum)))
+			return errors.New("inconsistent index values")
+		}
+	}
+	return nil
+}
+
+// extractRowMutations extracts row mutations and classify them into 2 categories: put and delete
+func extractRowMutations(mutations []mutation) ([]mutation, []mutation) {
```

Comment:
Could the classification be done in `collectTableMutationsFromBufferStage` so we could save one loop cost.

Reply:
Yep